log/logger_impl.go

```diff
@@ -0,0 +1,132 @@
+package log
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+)
+
+func NewDefaultLogger() Logger {
+	return NewLogger(log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr)))
+}
+
+func NewNopLogger() Logger {
+	return NewLogger(log.NewNopLogger())
+}
+
+func NewJSONLogger() Logger {
+	return NewLogger(log.NewJSONLogger(log.NewSyncWriter(os.Stderr)))
+}
+
+func NewBufferLogger() (*strings.Builder, Logger) {
+	buffer := strings.Builder{}
+	writer := log.NewLogfmtLogger(log.NewSyncWriter(&buffer))
+	log := NewLogger(writer)
+	return &buffer, log
+}
+
+func NewLogger(writer log.Logger) Logger {
+	l := &logger{
+		writer:  writer,
+		ctx:     map[string]string{},
+		lastErr: nil,
+	}
+
+	// Default logs to be info until changed
+	return l.Info()
+}
+
+type logger struct {
+	writer  log.Logger
+	ctx     map[string]string
+	lastErr error
+}
+
+func (l *logger) Set(key, value string) Logger {
+	return l.With(Fields{
+		key: value,
+	})
+}
+
+// With returns a new Logger with the contexts added to its own.
+func (l *logger) With(ctxs ...Context) Logger {
+	// Estimation assuming that for each ctxs has at least 1 value.
+	combined := make(map[string]string, len(l.ctx)+len(ctxs))
+
+	for k, v := range l.ctx {
+		combined[k] = v
+	}
+
+	for _, c := range ctxs {
+		itemCtx := c.Context()
+		for k, v := range itemCtx {
+			combined[k] = v
+		}
+	}
+
+	return &logger{
+		writer:  l.writer,
+		ctx:     combined,
+		lastErr: l.lastErr,
+	}
+}
+
+func (l *logger) WithError(err error) Logger {
```

Comment:
We shouldn't add this method since it forces the caller to set `"msg"` with a call to `Log()`.

For example, we should be able to do
```go
_, err := createCustomer()
if err != nil {
  err = logger.LogError(fmt.Errorf("creating customer: %v, err))
  // write err to resp body or whatever
}
```
which should output a log without `"msg"`
```json
{ 
"error": "...some error message..."
}
```

---- 
Using `WithError()` makes us do
```go
err = log.WithError(err).Log("message %w", err)
```
which duplicates information in the `"error"` and `"msg"` fields of the log.

Reply:
I'm a big fan of this

Reply:
The `.WithError` is such that we can return the exact error type along with logging it for `.LogError`. The intend is to what you mention is a rare case which I say should be used more. `msg` is to give context to the action being taken when it failed, while `err` is the failure itself.

For instance, I'm looking up an identity.
```
identity, err := repo.getIdentity(someID)
if err != nil {
  return nil, logger.WithError(err).Log("unable to find identity {someID}"))
}
```

The log output ion this case would be.
```
msg="unable to find identity 1324512345"
error="sql.NoRows"
```

msg is the context of the action being taken when ti failed.
error is the reasoning why it failed.

Having the error stored as the original type allows us to match it later on without string matching it. As not all errors are just a simple string type.

Reply:
In your example, wouldn't you want to add that context to the error? Such that the error would be:
`unable to find identity with {id}: sql: no rows in result set`.

In most cases where we're logging (the http handler level) you're not going to get the original error back, it'll hopefully be wrapped with additional context as it travels up the call stack.

Reply:
At the `http handler level` you need the original error types intact for it to do the switches on. You don't want to change it. So in your example we'd be switching the `sql.NoRows` error into a standard `error` type. When it gets to the http layer and inspects the type in your example it will return a `500` http status code because its no longer the `sql.NoRows` type to return a `404`

Reply:
That shouldn't be an issue using `errors.Is()`, `errors.As()`, or some other check.

Even on validations it'll be wrapped and output something like `invalid key: sql: dupliate key value`.